### PR TITLE
[Code] Tispark support insert using SQL

### PR DESCRIPTION
**Pull request by TrafalgarRicardoLu**
### What problem does this PR solve? 
https://github.com/pingcap/tispark/issues/2462

### What is changed and how it works?
- Add new capability to TiDBTable
- Remove write options
- Implement V2 interface using V1 code
- Compatible with different Spark versions
- close  https://github.com/pingcap/tispark/issues/2462
### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

### CODE REVIEW

File: docs/design/2022-07-20-insert-support.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
+# Support Insert SQL 
+
+- Author(s): TrafalgarRicardoLu
```
Comment (xuanyu66) [inline docs/design/2022-07-20-insert-support.md:3]: duplicated design doc？

File: core/src/main/scala/com/pingcap/tispark/write/TiDBOptions.scala [hunk 2]
```diff
@@ -135,7 +135,7 @@ class TiDBOptions(@transient val parameters: CaseInsensitiveMap[String]) extends
   // Optional parameters for read
   // ------------------------------------------------------------
   val tidbRowId: Boolean = getOrDefault(TIDB_ROWID, "false").toBoolean
-
+  val tidbRowFormatVersion: Int = getOrDefault(TiDB_ROW_FORMAT_VERSION, "2").toInt
```
Comment (xuanyu66) [inline core/src/main/scala/com/pingcap/tispark/write/TiDBOptions.scala:138]: It's better to write comment to explain why use 2 as default

**Review by xuanyu66 — APPROVED** [review]

**Review by xuanyu66 — CHANGES_REQUESTED** [review]
Plz add use doc for reviewing

File: docs/insert_sql_userguide.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+## Type Conversion for Write
+
+The following types of SparkSQL Data are currently not supported for writing to TiDB:
+
+- ArrayType
+- MapType
+- StructType
+
+Spark SQL and TIDB have different conversion rules. We follow the rule of TIDB when using DataSource API,
```
Comment (xuanyu66) [inline docs/insert_sql_userguide.md:86]: Plz give an example of how to cast such as String2TimeStamp and make spark link

File: core/src/main/scala/com/pingcap/tispark/v2/TiDBTable.scala [hunk 4]
```diff
@@ -128,21 +130,31 @@ case class TiDBTable(
...
   override def newWriteBuilder(info: LogicalWriteInfo): WriteBuilder = {
     var scalaMap = info.options().asScala.toMap
-    // TODO https://github.com/pingcap/tispark/issues/2269 we need to move TiDB dependencies which will block insert SQL.
-    // if we don't support it before release, insert SQL should throw exception in catalyst
-    if (scalaMap.isEmpty) {
-      throw new TiBatchWriteException("tidbOption is neccessary.")
+    // For update statics
+    val enableUpdateStatics =
+      sqlContext.getConf(TiDBOptions.TIDB_ENABLE_UPDATE_TABLE_STATISTICS, "false").toBoolean
+    if (enableUpdateStatics) {
```
Comment (shiyuhang0) [inline core/src/main/scala/com/pingcap/tispark/v2/TiDBTable.scala:136]: Just judge if the tidb config exists. if so, add them to scalaMap.
Judge by enableUpdateStatics is not a good idea. what if we need tidb config in another situation？

File: core/src/main/scala/com/pingcap/tispark/write/TiDBOptions.scala [hunk 5]
```diff
@@ -282,6 +286,7 @@ object TiDBOptions {
   val TIDB_ENABLE_UPDATE_TABLE_STATISTICS: String = newOption("enableUpdateTableStatistics")
   val TIDB_DEDUPLICATE: String = newOption("deduplicate")
   val TIDB_ROWID: String = newOption("tidbRowId")
+  val TiDB_ROW_FORMAT_VERSION: String = newOption("rowFormatVersion")
```
Comment (shiyuhang0) [inline core/src/main/scala/com/pingcap/tispark/write/TiDBOptions.scala:289]: how can insert sql pass this config?

File: core/src/test/scala/com/pingcap/tispark/insert/InsertSuite.scala [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.pingcap.tispark.insert
+
+import com.pingcap.tispark.datasource.BaseBatchWriteTest
+
+class InsertSuite extends BaseBatchWriteTest("test_delete_compatibility") {
```
Comment (shiyuhang0) [inline core/src/test/scala/com/pingcap/tispark/insert/InsertSuite.scala:21]: add test for `enableUpdateTableStatistics=true` and `rowFormatVersion=1`

File: spark-wrapper/spark-3.0/src/main/scala/org/apache/spark/sql/connector/write/TiDBWriteBuilder.scala [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+      override def insert(data: DataFrame, overwrite: Boolean): Unit = {
+        val schema = info.schema()
+        val df = sqlContext.sparkSession.createDataFrame(data.toJavaRDD, schema)
+        df.write
+          .format("tidb")
+          .options(tiDBOptions.parameters)
+          .option(TiDBOptions.TIDB_DATABASE, tiDBOptions.database)
+          .option(TiDBOptions.TIDB_TABLE, tiDBOptions.table)
+          .mode(SaveMode.Append)
+          .save()
```
Comment (shiyuhang0) [inline spark-wrapper/spark-3.0/src/main/scala/org/apache/spark/sql/connector/write/TiDBWriteBuilder.scala:40]: deduplicate needs to be set to false here, or it will have a different behavior from tidb

File: spark-wrapper/spark-3.1/src/main/scala/org/apache/spark/sql/connector/write/TiDBWriteBuilder.scala [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+      override def insert(data: DataFrame, overwrite: Boolean): Unit = {
+        val schema = info.schema()
+        val df = sqlContext.sparkSession.createDataFrame(data.toJavaRDD, schema)
+        df.write
+          .format("tidb")
+          .options(tiDBOptions.parameters)
+          .option(TiDBOptions.TIDB_DATABASE, tiDBOptions.database)
+          .option(TiDBOptions.TIDB_TABLE, tiDBOptions.table)
+          .mode(SaveMode.Append)
+          .save()
```
Comment (shiyuhang0) [inline spark-wrapper/spark-3.1/src/main/scala/org/apache/spark/sql/connector/write/TiDBWriteBuilder.scala:39]: ditto

File: spark-wrapper/spark-3.2/src/main/scala/org/apache/spark/sql/connector/write/TiDBWriteBuilder.scala [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+            val schema = info.schema()
+            println("Do write")
+            val df = sqlContext.sparkSession.createDataFrame(data.toJavaRDD, schema)
+            df.write
+              .format("tidb")
+              .options(tiDBOptions.parameters)
+              .option(TiDBOptions.TIDB_DATABASE, tiDBOptions.database)
+              .option(TiDBOptions.TIDB_TABLE, tiDBOptions.table)
+              .mode(SaveMode.Append)
+              .save()
```
Comment (shiyuhang0) [inline spark-wrapper/spark-3.2/src/main/scala/org/apache/spark/sql/connector/write/TiDBWriteBuilder.scala:42]: ditto

File: docs/insert_sql_userguide.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+          "insert into tpch_test.target_table_orders values" +
+            "(1,1,'0',11.11,date'2022-1-1','first','spark',12345,'no comment')")
+
+    // select data to write
+    val df = spark.sql("select * from tpch_test.ORDERS")
+   ```
+
+## TiDB Options
+
+The following table shows the TiDB-specific options, which can be passed in through `TiDBOptions` or `SparkConf`.
```
Comment (shiyuhang0) [inline docs/insert_sql_userguide.md:50]: Spark SQL can't accept TiDBOptions. So, we need not mention about`TiDBOptions` here?

File: docs/insert_sql_userguide.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+| tidb.password               | false    | -                      | TiDB Password, needed when enableUpdateTableStatistics is true                                                           |
+| enableRegionSplit           | false    | true                   | To split Region to avoid hot Region during insertion                                                                     |
+| scatterWaitMS               | false    | 300000                 | Max time to wait scatter region                                                                                          |
+| writeThreadPerTask          | false    | 1                      | Thread number each spark task use to write data to TiKV                                                                  |
+| bytesPerRegion              | false    | 100663296 (96M)        | Decrease this parameter to split more regions (increase write concurrency)                                               |
+| enableUpdateTableStatistics | false    | false                  | Update statistics in table `mysql.stats_meta` (`tidb.user` must own update privilege to table `mysql.stats_meta` if set true) |
```
Comment (shiyuhang0) [inline docs/insert_sql_userguide.md:65]: - I think you can't pass enableRegionSplit,writeThreadPerTask,bytesPerRegion,scatterWaitMS beacause you don't get them in `newWriteBuilder`.
- database , table and pd.addresses is needn't.
- lack TiDB_ROW_FORMAT_VERSION

File: docs/insert_sql_userguide.md (hunk 3 above)
Comment (shiyuhang0) [inline docs/insert_sql_userguide.md:86]: I think DataSource API may not follow TiDB

File: docs/insert_sql_userguide.md [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+## Type Conversion for Write
+
+The following types of SparkSQL Data are currently not supported for writing to TiDB:
+
+- ArrayType
+- MapType
+- StructType
+
+Spark SQL and TIDB have different conversion rules. We follow the rule of TIDB when using DataSource API,
+but follow the rule of Spark SQL when using it.
```
Comment (shiyuhang0) [inline docs/insert_sql_userguide.md:87]:
```suggestion
but follow the rule of Spark SQL when using insert.
```

File: core/src/main/scala/com/pingcap/tispark/write/TiDBOptions.scala (hunk 5 above)
Comment (TrafalgarRicardoLu) [reply core/src/main/scala/com/pingcap/tispark/write/TiDBOptions.scala:289]: users need to set it by `spark.conf.set()`. I'll add it  in `newWriteBuilder `

File: docs/insert_sql_userguide.md (hunk 11 above)
Comment (TrafalgarRicardoLu) [reply docs/insert_sql_userguide.md]: I have fix point two and three. For point one, we will get them in  newWriteBuilder now.

File: core/src/main/scala/com/pingcap/tispark/write/TiDBOptions.scala (hunk 5 above)
Comment (shiyuhang0) [reply core/src/main/scala/com/pingcap/tispark/write/TiDBOptions.scala:289]: yes. you need to convert any configs users may set to tidboption in newWriteBuilder. And these configs should be in your doc

File: docs/insert_sql_userguide.md [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+      `O_ORDERDATE` date NOT NULL,
+      `O_ORDERPRIORITY` char(15) NOT NULL,
+      `O_CLERK` char(15) NOT NULL,
+      `O_SHIPPRIORITY` int(11) NOT NULL,
+      `O_COMMENT` varchar(79) NOT NULL
+    )
+    */
+    
+   // write data to tidb
+    spark.sql(
```
Comment (zhangyangyu) [inline docs/insert_sql_userguide.md:40]: we don't need catalog?

File: docs/insert_sql_userguide.md [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+| tidb.addr                   | false    | -               | TiDB address, needed when enableUpdateTableStatistics is true                                                                 |
+| tidb.port                   | false    | -               | TiDB Port, needed when enableUpdateTableStatistics is true                                                                    |
+| tidb.user                   | false    | -               | TiDB User, needed when enableUpdateTableStatistics is true                                                                    |
+| tidb.password               | false    | -               | TiDB Password, needed when enableUpdateTableStatistics is true                                                                |
+| enableRegionSplit           | false    | true            | To split Region to avoid hot Region during insertion                                                                          |
+| scatterWaitMS               | false    | 300000          | Max time to wait scatter region                                                                                               |
```
Comment (zhangyangyu) [inline docs/insert_sql_userguide.md:59]: Why delete doesn't have so many options?

Comment (TrafalgarRicardoLu) [reply docs/insert_sql_userguide.md:59]: This options are used in write. Some of them are used to create jdbc client to update statistics, some of them are used to do region splits. 

Delete doesn't do these things. So it doesn't need.

File: docs/insert_sql_userguide.md (hunk 13 above)
Comment (TrafalgarRicardoLu) [reply docs/insert_sql_userguide.md:40]: I‘ll  add it.

File: docs/insert_sql_userguide.md [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+   //use catalog
+   spark.sql("use tidb_catalog")
+    
+   // write data to tidb
+    spark.sql(
+          "insert into tpch_test.target_table_orders values" +
+            "(1,1,'0',11.11,date'2022-1-1','first','spark',12345,'no comment')")
+
+    // select data
+    val df = spark.sql("select * from tpch_test.target_table_orders")
```
Comment (shiyuhang0) [inline docs/insert_sql_userguide.md:48]:
```suggestion
    spark.sql("select * from tpch_test.target_table_orders").show
```
use show to trigger an action,so that user can see the result

File: docs/insert_sql_userguide.md [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+   ```
+
+## TiDB Options
+
+The following table shows the TiDB-specific options, which can be passed in through `SparkConf`.
```
Comment (shiyuhang0) [inline docs/insert_sql_userguide.md:53]: password can't be passed by SparkConf, it can only be passed by `spark.conf.set`. they are different for the latter one is session level config, and it will include the configs in SparkConf (`SparkContext.getOrCreate().getConf`)
you need to remind user not set password in SparkConf at least. about other configs, they can set by both the SparkConf and `spark.conf.set`

File: docs/insert_sql_userguide.md [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+          "insert into tpch_test.target_table_orders values" +
+            "(1,1,'0',11.11,date'2022-1-1','first','spark',12345,'no comment')")
+
+    // select data
+    spark.sql("select * from tpch_test.target_table_orders").show
+   ```
+
+## TiDB Options
+
+The following table shows the TiDB-specific options, which can be passed in through `SparkConf` or `spark.conf.set`.
```
Comment (shiyuhang0) [inline docs/insert_sql_userguide.md:53]:
```suggestion
The following table shows the insert configurations, which can be passed in through `SparkConf` or `spark.conf.set`.
```

File: docs/insert_sql_userguide.md [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+   // write data to tidb
+    spark.sql(
+          "insert into tpch_test.target_table_orders values" +
+            "(1,1,'0',11.11,date'2022-1-1','first','spark',12345,'no comment')")
+
+    // select data
+    spark.sql("select * from tpch_test.target_table_orders").show
+   ```
+
+## TiDB Options
```
[the remark this answers is not in the document]
Comment (shiyuhang0) [reply docs/insert_sql_userguide.md:51]: I think options refer to datasource options. it is better to use configs or Configurations  in spark sql

**Review by shiyuhang0 — APPROVED** [review]

File: core/src/main/scala/com/pingcap/tispark/write/TiDBOptions.scala (hunk 5 above)
Comment (qidi1) [inline core/src/main/scala/com/pingcap/tispark/write/TiDBOptions.scala:289]: Why a new mechanism was introduced instead of following the past practice?https://github.com/pingcap/tispark/blob/cc9c121dab6dff68fbb7ca1cae4c6b662845c7b8/core/src/main/scala/com/pingcap/tispark/write/TiBatchWriteTable.scala#L55-L56

Comment (TrafalgarRicardoLu) [reply core/src/main/scala/com/pingcap/tispark/write/TiDBOptions.scala:289]: we decide to remove jdbc client and only use it to update statistics. This option was gotten by jdbc client, so we are not able to get it now. Instead, we put it in TiDBOption and users can set it by `spark.conf.set`.

File: spark-wrapper/spark-3.0/src/main/scala/org/apache/spark/sql/connector/write/TiDBWriteBuilder.scala [hunk 19]
```diff
@@ -36,6 +36,7 @@ case class TiDBWriteBuilder(
           .options(tiDBOptions.parameters)
           .option(TiDBOptions.TIDB_DATABASE, tiDBOptions.database)
           .option(TiDBOptions.TIDB_TABLE, tiDBOptions.table)
+          .option(TiDBOptions.TIDB_DEDUPLICATE, "false")
```
Comment (xuanyu66) [inline spark-wrapper/spark-3.0/src/main/scala/org/apache/spark/sql/connector/write/TiDBWriteBuilder.scala:39]: Why add this option, is it a bug fix?

Comment (TrafalgarRicardoLu) [reply spark-wrapper/spark-3.0/src/main/scala/org/apache/spark/sql/connector/write/TiDBWriteBuilder.scala:39]: We want

Comment (TrafalgarRicardoLu) [reply spark-wrapper/spark-3.0/src/main/scala/org/apache/spark/sql/connector/write/TiDBWriteBuilder.scala:39]: We want to have same behavior with TiDB.  TiDB will throw an error when there are duplicate keys. If it's `true`, we will handle it instead of throw an error.

**Review by qidi1 — APPROVED** [review]

**Review by xuanyu66 — APPROVED** [review]

**Review by shiyuhang0 — APPROVED** [review]

**Review by xuanyu66 — APPROVED** [review]

File: core/src/main/java/com/pingcap/tikv/datatype/TypeMapping.java [hunk 20]
```diff
@@ -90,6 +80,9 @@ public static org.apache.spark.sql.types.DataType toSparkType(DataType type) {
       if (type.isUnsigned() && type.getType() == TypeLonglong) {
         return DataTypes.createDecimalType(20, 0);
       }
+      if (type.getType() == TypeTiny && type.getLength() == 1) {
```
Comment (boneanxs) [inline core/src/main/java/com/pingcap/tikv/datatype/TypeMapping.java:83]: Hey @TrafalgarRicardoLu It appeals this is a breaking change, can explain more why this change is required?

Comment (zhangyangyu) [reply core/src/main/java/com/pingcap/tikv/datatype/TypeMapping.java:83]: I think no. IIRC MySQL doesn't have boolean type before, so everyone use 1 bit tinyint to represent a boolean type. Many MySQL ecosystem softwares still streat this as bool.

Comment (boneanxs) [reply core/src/main/java/com/pingcap/tikv/datatype/TypeMapping.java:83]: Yea, understand this is trying to align with others databases, but for someone who rely on this before, after upgrading to the new version, they possibly meet issues. Just an advice, maybe we can communicate such changes in the changelog to alert users who relied on the previous behavior.

Comment (zhangyangyu) [reply core/src/main/java/com/pingcap/tikv/datatype/TypeMapping.java:83]: Yeah, reasonable. We could do better. Does this change affect you?

Comment (boneanxs) [reply core/src/main/java/com/pingcap/tikv/datatype/TypeMapping.java:83]: Yea, we actually use this to ingest data from TiDB to lakehouse through HUDI, and because of this schema change, the target table could meet schema incompatible issues.
